File: 2021/Day 10/Part 1/day10_1.cpp
#include <iostream>
#include <stack>
#include <unordered_map>

using namespace std;

const unordered_map<char, char> pairs = { {')', '('}, {']', '['}, {'}', '{'}, {'>', '<'} };
const unordered_map<char, int> scores = { {')', 3}, {']', 57}, {'}', 1197}, {'>', 25137} };

int getErrorScore(string& line) {
    stack<char> stack;

    for (char ch : line) {
        if (ch == '(' || ch == '[' || ch == '{' || ch == '<') {
            stack.push(ch);
        } else if (stack.top() == pairs.at(ch)) {
            stack.pop();
        } else {
            return scores.at(ch);
        }
    }

    return 0;
}

int main() {
    int sum = 0;

    string line;
    while (cin >> line) {
        sum += getErrorScore(line);
    }

    cout << sum;

    return 0;
}

File: 2021/Day 10/Part 1/Day10_1.java
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;
import java.util.Stack;

public class Day10_1 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        Map<Character, Integer> syntaxCheckerScores = new HashMap<>();
        syntaxCheckerScores.put(')', 3);
        syntaxCheckerScores.put(']', 57);
        syntaxCheckerScores.put('}', 1197);
        syntaxCheckerScores.put('>', 25137);

        int sum = 0;

        while (sc.hasNextLine()) {
            char[] line = sc.nextLine().toCharArray();
            sum += getErrorScore(line, syntaxCheckerScores);
        }

        System.out.println(sum);
        sc.close();
    }

    private static int getErrorScore(char[] line, Map<Character, Integer> syntaxCheckerScores) {
        Stack<Character> stack = new Stack<>();

        for (char character : line) {
            switch (character) {
                case '(', '[', '{', '<' -> stack.push(character);
                case ')' -> { if (stack.pop() != '(') return syntaxCheckerScores.get(character); }
                case ']' -> { if (stack.pop() != '[') return syntaxCheckerScores.get(character); }
                case '}' -> { if (stack.pop() != '{') return syntaxCheckerScores.get(character); }
                case '>' -> { if (stack.pop() != '<') return syntaxCheckerScores.get(character); }
            }
        }

        return 0;
    }
}
